docs/guide/api-vite-environment.md

```diff
@@ -0,0 +1,925 @@
+  return new DevEnvironment(name, config, {
+    hot: /* custom hot channel */,
+    runner: {
+      transport: new RemoteEnvironmentTransport({
+        send: (data) => worker.postMessage(data),
+        onMessage: (listener) => worker.on('message', listener),
+      }),
+    },
+  })
+}
+
+await createServer({
+  environments: {
+    worker: {
+      dev: {
+        createEnvironment: createWorkerEnvironment,
+      },
+    },
+  },
+})
+```
+
+:::
+
+`RemoteRunnerTransport` and `RemoteEnvironmentTransport` are meant to be used together. If you don't use either of them, then you can define your own function to communicate between the runner and the server.
+
+## ModuleRunnerHMRConnection
+
+**Type Signature:**
+
+```ts
+export interface ModuleRunnerHMRConnection {
+  /**
+   * Checked before sending messages to the client.
+   */
+  isReady(): boolean
+  /**
+   * Send a message to the client.
+   */
+  send(message: string): void
+  /**
+   * Configure how HMR is handled when this connection triggers an update.
+   * This method expects that the connection will start listening for HMR updates and call this callback when it's received.
+   */
+  onUpdate(callback: (payload: HotPayload) => void): void
+}
+```
+
+This interface defines how HMR communication is established. Vite exports `ServerHMRConnector` from the main entry point to support HMR during Vite SSR. The `isReady` and `send` methods are usually called when the custom event is triggered (like, `import.meta.hot.send("my-event")`).
+
+`onUpdate` is called only once when the new module runner is initiated. It passed down a method that should be called when connection triggers the HMR event. The implementation depends on the type of connection (as an example, it can be `WebSocket`/`EventEmitter`/`MessageChannel`), but it usually looks something like this:
+
+```js
+function onUpdate(callback) {
+  this.connection.on('hmr', (event) => callback(event.data))
+}
+```
+
+The callback is queued and it will wait for the current update to be resolved before processing the next update. Unlike the browser implementation, HMR updates in a module runner will wait until all listeners (like, `vite:beforeUpdate`/`vite:beforeFullReload`) are finished before updating the modules.
+
+## Environments during build
+
+In the CLI, calling `vite build` and `vite build --ssr` will still build the client only and ssr only environments for backward compatibility.
+
+When `builder.entireApp` is `true` (or when calling `vite build --app`), `vite build` will opt-in into building the entire app instead. This would later on become the default in a future major. A `ViteBuilder` instance will be created (build-time equivalent to a `ViteDevServer`) to build all configured environments for production. By default the build of environments is run in series respecting the order of the `environments` record. A framework or user can further configure how the environments are built using:
```

Comment:
We moved to `build --app` and `builder.buildApp` (instead of `--all` and `builder.buildEnvironments`), because frameworks are probably going to use `buildApp` to do more steps than just build the environments there (copy extra files, fire other tools, etc). And because we wanted to also build libraries as part of this step (in case we allow for multiple libraries, like an environment with libmode true). I checked and we were only using `all` in ssr-noexternal.